File: output_20260319_145930/claude-opus/c/Struct/spouse.c
#include "stdlib.h"
#include "spouse.h"

struct person {
  char* name;
  struct person* spouse;
};

/*@
predicate person(struct person *p;) =
  p->name |-> _ &*& p->spouse |-> ?s &*&
  (s == 0 ? true : s->spouse |-> p);
@*/

/*@
predicate person_single(struct person *p;) =
  p->name |-> _ &*& p->spouse |-> 0;
@*/

struct person *create_person()
  //@ requires true;
  //@ ensures person_single(result);
{
  struct person *p = malloc(sizeof(struct person));
  if(p == 0) abort();
  p->spouse = 0;
  //@ close person_single(p);
  return p;
}

void marry(struct person *this, struct person *other)
  //@ requires person_single(this) &*& person_single(other) &*& this != other;
  //@ ensures person(this) &*& this->spouse |-> other;
{
  //@ open person_single(this);
  //@ open person_single(other);
  this->spouse = other;
  other->spouse = this;
  //@ close person(this);
}

struct person* person_get_spouse(struct person* this)
  //@ requires this->spouse |-> ?s;
  //@ ensures this->spouse |-> s &*& result == s;
{
  return this->spouse;
}

void divorce(struct person* this)
  //@ requires person(this) &*& this->spouse |-> ?other &*& other != 0;
  //@ ensures person_single(this) &*& person_single(other);
{
  //@ open person(this);
  this->spouse->spouse = 0;
  //@ close person_single(this->spouse);
  this->spouse = 0;
  //@ close person_single(this);
}

void die(struct person *this)
  //@ requires person_single(this) &*& (this->spouse |-> 0 ? true : true);
  //@ ensures true;
{
  //@ open person_single(this);
  if(this->spouse != 0) {
    this->spouse->spouse = 0;
  }
  free(this); 
}

int main() 
  //@ requires true;
  //@ ensures true;
{
  struct person* alice = create_person();
  struct person* bob = create_person();
  struct person* eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  //@ open person(alice);
  die(alice);
  //@ open person(bob);
  die(bob);
  die(eve);
  return 0;
}

File: output_20260320_065752/deepseek/java/Linklist/Spouse.java
class Person {

    protected Person spouse;

    /*@
    predicate Person() = spouse |-> ?s;
    @*/

    //@ requires true;
    //@ ensures Person();
    public  void spouse_symm()
        //@ requires Person();
        //@ ensures Person();
    {
        //@ open Person();
        //@ close Person();
    }

    //@ ensures Person();
    public Person()
        //@ ensures Person();
    {
        //@ close Person();
    }
    
    //@ requires Person();
    //@ ensures Person() &*& result == spouse;
    public Person getSpouse()
        //@ requires Person();
        //@ ensures Person() &*& result == spouse;
    {
        //@ open Person();
        //@ close Person();
        return spouse;
    }
    
    //@ requires Person() &*& other != null &*& other.Person();
    //@ ensures Person() &*& other.Person();
    protected void setSpouse(Person other)
        //@ requires Person() &*& other != null &*& other.Person();
        //@ ensures Person() &*& other.Person();
    {
        //@ open Person();
        //@ open other.Person();
        spouse = other;
        other.spouse = this;
        //@ close other.Person();
        //@ close Person();
    }
    
    //@ requires Person() &*& spouse != null &*& spouse.Person();
    //@ ensures Person();
    protected void clearSpouse()
        //@ requires Person() &*& spouse != null &*& spouse.Person();
        //@ ensures Person();
    {
        //@ open Person();
        //@ open spouse.Person();
        spouse.spouse = null;
        spouse = null;
        //@ close Person();
    }
    
    //@ requires Person() &*& other != null &*& other.Person();
    //@ ensures Person() &*& other.Person();
    void marry(Person other)
        //@ requires Person() &*& other != null &*& other.Person();
        //@ ensures Person() &*& other.Person();
    {
        other.setSpouse(this);
    }
    
    //@ requires Person() &*& spouse != null &*& spouse.Person();
    //@ ensures Person();
    void divorce()
        //@ requires Person() &*& spouse != null &*& spouse.Person();
        //@ ensures Person();
    {
        spouse.clearSpouse();
    }

}

class Program {

    public static void main(String[] args)
        //@ requires true;
        //@ ensures true;
    {
        Person a = new Person();
        Person b = new Person();
        a.marry(b);
        b.divorce();
    }

}